### DISCUSSION TITLE
Set a minimum PHP/Roundcube version

### DISCUSSION THREAD

**Discussion by williamdes** [GermanCoding/Roundcube_TLS_Icon#6] [Q&A]
To improve coding this extension I would like to set a minimum roundcube (PHP version) for this plugin.
What do you think is reasonable viewing: https://github.com/roundcube/roundcubemail/wiki/Version-History and the installs by PHP version stats: https://packagist.org/packages/germancoding/tls_icon/php-stats ?

The benefits are that when this is defined I can start requiring some tools (like phpunit) without having a too large PHP version to handle.

In my opinion we should require PHP 5.6 since the install stats show it is used some times. And that allowing people to use this plugin on a 2017 Roundcube version is of not use to anyone. So I would like to require at a minimum a 2019 version: 1.4

So it seems like this would point to requiring: `>= 1.4.x` and so `PHP >= 5.4`

Feel free to take this to another level of requiring because of composer dropping old PHP versions:
- https://blog.packagist.com/composer-2-0-is-now-available/
- https://blog.packagist.com/deprecating-composer-1-support/

**Comment by GermanCoding** [comment]
Well on one hand I have little mercy with people running software that hasn't seen a security update in years. On the other hand I also don't like creating needless obstacles by requring modern versions without a real benefit.

In the end, this all boils down to "how much pain is it to support old version X.Y?"
-> If easy: Let's support it.
-> If painful: Drop it.

In an ideal world, we would just support whatever Roundcube does. Roundcube 1.3 and 1.4 are _technically_ still in LTS support and require at least PHP 5.4, but is there really anyone still supplying PHP 5.4 with actual fixes?

Supporting only Roundcube 1.4 and up is fine to me, if Roundcube 1.3 seems to much of a hassle. I personally don't even run an internal test deployment for 1.3 anymore, and am going to decomission my 1.4 tests soon (because of 1.6 right around the corner).

As for composer, last time I checked the roundcube plugin installer could work with composer 1 or 2 (though one had to pick the right version). I remember having a lot of issues migrating to composer 2 initially though, but these days my composer 2 installers seem to work without too many issues. I think it would be preferable if we could remain compatible with old roundcube plugin installers/composer 1, as there are probably a lot of images out there still using it. At least for the time being, eventually we will all be running on composer 2. According to your links that will also bring a version bump to PHP ~7, so at that point it hardly makes sense to support older PHP versions too.

**Comment by williamdes** [comment]
> Supporting only Roundcube 1.4 and up is fine to me, if Roundcube 1.3 seems to much of a hassle. I personally don't even run an internal test deployment for 1.3 anymore, and am going to decomission my 1.4 tests soon (because of 1.6 right around the corner).

1.3 or 1.4 have the same minimum PHP requirements so I think there is no effort to support 1.3
And since 5.4 has the new array syntax that's good for cleaning a bit the code up: https://github.com/wdes/coding-standard#rules-to-disable-for-php--54-compatibility

And maybe when 1.6 is out for some time we can bump the requirements a bit but that will not help much with coding here ^^
Let's make some years pass and bump again 😄 

I think we can conclude on my next PR where I will require 1.3 and PHP 5.4 as minimums

**Comment by williamdes** [comment]
You now what, this plugin is not compatible with PHP 5 because of the `??` in the code, fixed in https://github.com/GermanCoding/Roundcube_TLS_Icon/pull/7
See: https://www.php.net/manual/en/migration70.new-features.php#migration70.new-features.null-coalesce-op
